### PR TITLE
pandoc: remove reference to warp

### PR DESCRIPTION
**Pull request by berberman**
###### Description of changes

Current closure size of `pandoc` is about 3G, as GHC is included:

```
$ nix why-depends nixpkgs#pandoc nixpkgs#ghc
/nix/store/0nczqyg5a8iglvj6gz64v8qigv2pnriy-pandoc-2.19.2
└───/nix/store/px1i2jfmnzrc6b3m6k8swwkv66bxzs0i-warp-3.3.23
    └───/nix/store/322zfsd89g1xph46glm0jjvwjkj09mv4-ghc-9.2.6
```

It's likely that we are in the situation similar to #34376, but for `warp` this time, given that `wrap` also uses Paths_* module for obtaining its version. Let's remove this reference.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by tobim** [comment]
This is great, but did you get any indication whether this breaks some functionality of pandoc?

File: pkgs/development/tools/pandoc/default.nix [hunk 1]
```diff
@@ -13,6 +13,9 @@ in
       remove-references-to \
         -t ${haskellPackages.pandoc-types} \
         $out/bin/pandoc
+      remove-references-to \
+        -t ${haskellPackages.warp} \
+        $out/bin/pandoc
```
Comment (Atemu) [inline pkgs/development/tools/pandoc/default.nix:18]: Why does pandoc depend on references to warp to begin with? Please add the reason this is safe to do as a comment.

**Review by Atemu — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/pandoc/default.nix (hunk 1 above)
Comment (berberman) [reply pkgs/development/tools/pandoc/default.nix:18]: By using `justStaticExecutables foo`,  the compiled executable shouldn't refer to any Haskell libraries and documentation. However, this is not always the case when the dependency has `Paths_` module - `bindir`, `datadir`, and `libdir` which contain the path to that Haskell library will still appear in the executable (https://github.com/NixOS/nixpkgs/issues/34376#issuecomment-375745951). Most Haskell programs only use `Paths_foo.version` for getting its version at runtime, while not using those directories, so it's safe to remove the reference to the Haskell library. And this is exactly the case of `warp`:

```haskell
Network/Wai/Handler/Warp/Settings.hs
21:import qualified Paths_warp
195:    , settingsServerName = C8.pack $ "Warp/" ++ showVersion Paths_warp.version

Network/Wai/Handler/Warp/Response.hs
34:import qualified Paths_warp
409:warpVersion = showVersion Paths_warp.version
```

**Comment by berberman** [comment]
> This is great, but did you get any indication whether this breaks some functionality of pandoc?

Haskell library `wrap` drives `pandoc-server`, and I tested it works fine.

File: pkgs/development/tools/pandoc/default.nix (hunk 1 above)
Comment (berberman) [reply pkgs/development/tools/pandoc/default.nix:18]: And there is already a comment at line 25 that explains everything.

Comment (Atemu) [reply pkgs/development/tools/pandoc/default.nix:18]: Could you copy paste your reply 1:1 into a comment? Context like this helps a lot when someone looks at this again in the future.

**Comment by GabrielDougherty** [comment]
Result of `nixpkgs-review pr 219931` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>43 packages marked as broken and skipped:</summary>
  <ul>
    <li>cardboard</li>
    <li>getoptions</li>
    <li>ghostwriter</li>
    <li>gssdp-tools</li>
    <li>gssdp_1_6</li>
    <li>gupnp-tools</li>
    <li>manim</li>
    <li>napari</li>
    <li>notcurses</li>
    <li>openvino</li>
    <li>python310Packages.Nikola</li>
    <li>python310Packages.arviz</li>
    <li>python310Packages.augmax</li>
    <li>python310Packages.bambi</li>
    <li>python310Packages.batchspawner</li>
    <li>python310Packages.dalle-mini</li>
    <li>python310Packages.dockerspawner</li>
    <li>python310Packages.elegy</li>
    <li>python310Packages.jupyterhub</li>
    <li>python310Packages.jupyterhub-ldapauthenticator</li>
    <li>python310Packages.jupyterhub-systemdspawner</li>
    <li>python310Packages.jupyterhub-tmpauthenticator</li>
    <li>python310Packages.n3fit</li>
    <li>python310Packages.napari</li>
    <li>python310Packages.napari-npe2</li>
    <li>python310Packages.oauthenticator</li>
    <li>python310Packages.openvino</li>
    <li>python310Packages.pymc</li>
    <li>python310Packages.treex</li>
    <li>python310Packages.validphys2</li>
    <li>python311Packages.Nikola</li>
    <li>python311Packages.augmax</li>
    <li>python311Packages.batchspawner</li>
    <li>python311Packages.dockerspawner</li>
    <li>python311Packages.jupyterhub</li>
    <li>python311Packages.jupyterhub-ldapauthenticator</li>
    <li>python311Packages.jupyterhub-systemdspawner</li>
    <li>python311Packages.jupyterhub-tmpauthenticator</li>
    <li>python311Packages.oauthenticator</li>
    <li>python311Packages.openvino</li>
    <li>python311Packages.validphys2</li>
    <li>run-scaled</li>
    <li>validphys2</li>
  </ul>
</details>
<details>
  <summary>11 packages failed to build:</summary>
  <ul>
    <li>jupyter</li>
    <li>python310Packages.ansible-kernel</li>
    <li>python310Packages.jupyter-server-mathjax</li>
    <li>python310Packages.jupyterlab-git</li>
    <li>python310Packages.nbdime</li>
    <li>python311Packages.ansible-kernel</li>
    <li>python311Packages.httpie</li>
    <li>python311Packages.httpie-ntlm</li>
    <li>python311Packages.jupyter-server-mathjax</li>
    <li>python311Packages.jupyterlab-git</li>
    <li>python311Packages.nbdime</li>
  </ul>
</details>
<details>
  <summary>95 packages built:</summary>
  <ul>
    <li>actionlint</li>
    <li>apksigcopier</li>
    <li>bashblog</li>
    <li>bup</li>
    <li>coursera-dl</li>
    <li>dogdns</li>
    <li>eget</li>
    <li>emanote</li>
    <li>exa</li>
    <li>fend</li>
    <li>gg-scm</li>
    <li>git-annex (haskellPackages.git-annex)</li>
    <li>git-ftp</li>
    <li>git-open</li>
    <li>gocryptfs</li>
    <li>haskellPackages.digestive-functors-heist</li>
    <li>haskellPackages.emanote</li>
    <li>haskellPackages.hackage-cli</li>
    <li>haskellPackages.heist</li>
    <li>haskellPackages.heist-extra</li>
    <li>haskellPackages.http-io-streams</li>
    <li>haskellPackages.serversession-frontend-snap</li>
    <li>haskellPackages.snap</li>
    <li>haskellPackages.snaplet-ghcjs</li>
    <li>haskellPackages.snaplet-i18n</li>
    <li>haskellPackages.snaplet-persistent</li>
    <li>haskellPackages.snaplet-purescript</li>
    <li>haskellPackages.snaplet-sqlite-simple</li>
    <li>haskellPackages.web-fpco</li>
    <li>http-prompt</li>
    <li>httpie (python310Packages.httpie)</li>
    <li>jo</li>
    <li>lsd</li>
    <li>memo</li>
    <li>mp3fs</li>
    <li>mpdcron</li>
    <li>nix-info-tested</li>
    <li>nominatim</li>
    <li>ov</li>
    <li>pandoc</li>
    <li>passExtensions.pass-checkup</li>
    <li>python310Packages.einops</li>
    <li>python310Packages.git-annex-adapter</li>
    <li>python310Packages.httpie-ntlm</li>
    <li>python310Packages.jupyter</li>
    <li>python310Packages.jupyter-lsp</li>
    <li>python310Packages.jupyter-server</li>
    <li>python310Packages.jupyterlab</li>
    <li>python310Packages.jupyterlab-lsp</li>
    <li>python310Packages.jupyterlab_launcher</li>
    <li>python310Packages.jupyterlab_server</li>
    <li>python310Packages.jupytext</li>
    <li>python310Packages.mkdocs-jupyter</li>
    <li>python310Packages.nbclassic</li>
    <li>python310Packages.nix-kernel</li>
    <li>python310Packages.notebook</li>
    <li>python310Packages.notebook-shim</li>
    <li>python310Packages.notedown</li>
    <li>python310Packages.pelican</li>
    <li>python310Packages.phik</li>
    <li>python310Packages.pymoo</li>
    <li>python310Packages.pypandoc</li>
    <li>python310Packages.pytm</li>
    <li>python310Packages.pytomlpp</li>
    <li>python310Packages.reportengine</li>
    <li>python310Packages.vector</li>
    <li>python310Packages.vega</li>
    <li>python310Packages.xarray-einstats</li>
    <li>python311Packages.einops</li>
    <li>python311Packages.git-annex-adapter</li>
    <li>python311Packages.jupyter</li>
    <li>python311Packages.jupyter-lsp</li>
    <li>python311Packages.jupyter-server</li>
    <li>python311Packages.jupyterlab</li>
    <li>python311Packages.jupyterlab-lsp</li>
    <li>python311Packages.jupyterlab_launcher</li>
    <li>python311Packages.jupyterlab_server</li>
    <li>python311Packages.jupytext</li>
    <li>python311Packages.mkdocs-jupyter</li>
    <li>python311Packages.nbclassic</li>
    <li>python311Packages.nix-kernel</li>
    <li>python311Packages.notebook</li>
    <li>python311Packages.notebook-shim</li>
    <li>python311Packages.notedown</li>
    <li>python311Packages.pelican</li>
    <li>python311Packages.pypandoc</li>
    <li>python311Packages.pytm</li>
    <li>python311Packages.pytomlpp</li>
    <li>python311Packages.reportengine</li>
    <li>python311Packages.vega</li>
    <li>ripgrep-all</li>
    <li>scriv</li>
    <li>shellcheck</li>
    <li>vimPlugins.cmp-pandoc-nvim</li>
    <li>wayback-machine-archiver</li>
  </ul>
</details>

**Review by Atemu — APPROVED** [review]
Result of `nixpkgs-review pr 219931` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>pandoc</li>
  </ul>
</details>

File: pkgs/development/tools/pandoc/default.nix [hunk 2]
```diff
@@ -9,10 +9,20 @@ in
     buildDepends = drv.buildDepends or [] ++ [haskellPackages.file-embed];
     buildTools = (drv.buildTools or []) ++ [ removeReferencesTo ];
 
+    # Normally, the static linked executable shouldn't refer to any library or the compiler.
+    # This is not always the case when the dependency has Paths_* module generated by Cabal,
+    # where bindir, datadir, and libdir contain the path to the library, and thus make the
+    # executable indirectly refer to GHC. However, most Haskell programs only use Paths_*.version for
+    # getting the version at runtime, so it's safe to remove the references to them.
+    # This is true so far for pandoc-types and warp.
```
Comment (sternenseemann) [inline pkgs/development/tools/pandoc/default.nix:17]: Confirmed: https://github.com/search?q=repo%3Ayesodweb%2Fwai%20Paths_warp&type=code

**Review by sternenseemann — APPROVED** [review]

**Comment by sternenseemann** [comment]
I'd say it is safer to let this one go via staging as the rebuild on linux is kind of hefty.

**Comment by Atemu** [comment]
> the rebuild on linux is kind of hefty.

Depends on what kind of package those are.

I had a quick look and at first glance, no "big" packages I know of seem to be affected. The biggest ones are probably the python TF packages.

Given that it's just 500 packages/platform (none of which seem particularly large to me), I don't think this needs to go through staging.

**Comment by GabrielDougherty** [comment]
I've been building on my laptop for about 14 hours and I am at `1/285/596` built. The packages I remember taking a long time were openvino, nccl-2.16.5-1-cuda-11, and ceph

This is just from running `nix run nixpkgs#nixpkgs-review pr 219931`

**Comment by berberman** [comment]
Let's make this land through staging anyway.

**Pull request merged**